File: leetcode/algorithms/706_design_hash_map/main.js
/**
 * Your MyHashMap object will be instantiated and called as such:
 * var obj = new MyHashMap()
 * obj.put(key,value)
 * var param_2 = obj.get(key)
 * obj.remove(key)
 */

var MyHashMap = function () {
  this.nums = Array(1e6 + 1).fill(-1);
};

/**
 * @param {number} key
 * @param {number} value
 * @return {void}
 */
MyHashMap.prototype.put = function (key, value) {
  this.nums[key] = value;
};

/**
 * @param {number} key
 * @return {number}
 */
MyHashMap.prototype.get = function (key) {
  return this.nums[key];
};

/**
 * @param {number} key
 * @return {void}
 */
MyHashMap.prototype.remove = function (key) {
  this.nums[key] = -1;
};


// Solution
// Solution 1: Array
class Solution1_MyHashMap {
  constructor() {
    this.data = new Array(1000001);
  }

  put(key, val) {
    this.data[key] = val;
  }

  get(key) {
    let val = this.data[key];
    return val !== undefined ? val : -1;
  }

  remove(key) {
    delete this.data[key];
  }
}

// Solution 2: Hash
class ListNode {
    constructor(key, val, next) {
        this.key = key;
        this.val = val;
        this.next = next;
    }
}

class Solution2_MyHashMap {
    constructor() {
        this.size = 19997;
        this.mult = 12582917;
        this.data = new Array(this.size);
    }

    hash(key) {
        return (key * this.mult) % this.size;
    }

    put(key, val) {
        this.remove(key);
        let h = this.hash(key);
        let node = new ListNode(key, val, this.data[h]);
        this.data[h] = node;
    }

    get(key) {
        let h = this.hash(key), node = this.data[h];

        for (; node; node = node.next) {
            if (node.key === key) {
                return node.val;
            }
        }

        return -1;
    }

    remove(key) {
        let h = this.hash(key), node = this.data[h];

        if (!node) {
            return;
        }

        if (node.key === key) {
            this.data[h] = node.next;
        } else {
            for (; node.next; node = node.next) {
                if (node.next.key === key) {
                    node.next = node.next.next;
                    return;
                }
            }
        }
    }
}

File: leetcode/algorithms/706_design_hash_map/main.java
import java.util.Arrays;

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap obj = new MyHashMap();
 * obj.put(key,value);
 * int param_2 = obj.get(key);
 * obj.remove(key);
 */

class MyHashMap {
    private final int[] nums = new int[1000001];

    public MyHashMap() {
        Arrays.fill(this.nums, -1);
    }

    public void put(int key, int value) {
        this.nums[key] = value;
    }

    public int get(int key) {
        return this.nums[key];
    }

    public void remove(int key) {
        this.nums[key] = -1;
    }
}


// Solution
// Solution 1: Array
class Solution1_MyHashMap {
    int[] data;

    public Solution1_MyHashMap() {
        data = new int[1000001];
        Arrays.fill(data, -1);
    }

    public void put(int key, int val) {
        data[key] = val;
    }

    public int get(int key) {
        return data[key];
    }

    public void remove(int key) {
        data[key] = -1;
    }
}

// Solution 2: Hash
class ListNode {
    int key, val;
    ListNode next;

    public ListNode(int key, int val, ListNode next) {
        this.key = key;
        this.val = val;
        this.next = next;
    }
}

class Solution2_MyHashMap {
    static final int size = 19997;
    static final int mult = 12582917;
    ListNode[] data;

    public Solution2_MyHashMap() {
        this.data = new ListNode[size];
    }

    private int hash(int key) {
        return (int) ((long) key * mult % size);
    }

    public void put(int key, int val) {
        remove(key);
        int h = hash(key);
        ListNode node = new ListNode(key, val, data[h]);
        data[h] = node;
    }

    public int get(int key) {
        int h = hash(key);
        ListNode node = data[h];

        for (; node != null; node = node.next) {
            if (node.key == key) {
                return node.val;
            }
        }

        return -1;
    }

    public void remove(int key) {
        int h = hash(key);
        ListNode node = data[h];

        if (node == null) {
            return;
        }

        if (node.key == key) {
            data[h] = node.next;
        } else {
            for (; node.next != null; node = node.next) {
                if (node.next.key == key) {
                    node.next = node.next.next;
                    return;
                }
            }
        }
    }
}
